core/compiler.common.jvm/src/org/jetbrains/kotlin/load/java/AbstractAnnotationTypeQualifierResolver.kt

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.load.java
+
+import org.jetbrains.kotlin.builtins.StandardNames
+import org.jetbrains.kotlin.descriptors.annotations.KotlinTarget
+import org.jetbrains.kotlin.name.FqName
+import java.util.concurrent.ConcurrentHashMap
+
+typealias TypeQualifierWithApplicability<Annotation> = Pair<Annotation, Set<AnnotationQualifierApplicabilityType>>
+
+abstract class AbstractAnnotationTypeQualifierResolver<Annotation : Any>(
+    private val javaTypeEnhancementState: JavaTypeEnhancementState
+) {
+    protected abstract val Annotation.annotations: Iterable<Annotation>
+    protected abstract val Annotation.key: Any
```

Comment:
Looks like using type constructors here would be equivalent to using fqnames, and I didn't do that because I'm not 100% sure that given two `ClassDescriptor`s for the same type we can use one's annotations for the other (e.g. because they're technically compiled as different modules in a daemon configuration or something?)

Reply:
Added a TODO to figure out if fqNames are fine as keys.